### PR TITLE
HOCS-2452 Add User

### PR DESCRIPTION
**Pull request by SeanReedHomeOffice**
1. New screens for adding and amending users.
2. A new checkbox component was introduced.
3. Validation messages from server side or client side can now be shown.
4. The existing User model object renamed to ListUser, to free up User for storing actual user data.

### CODE REVIEW

File: src/shared/models/constants.ts [hunk 1]
```diff
@@ -21,8 +25,9 @@ export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the
 export const ADD_UNIT_SUCCESS = 'The unit was added successfully';
 export const ADD_USER_TO_TEAM_SUCCESS = 'The user was added successfully';
 export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
-export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
-export const ADD_USER_SUCCESS = 'User(s) added successfully';
+export const ADD_USER_ERROR_TITLE = 'Add User Failed Due to server error';
```
Comment (zuzak) [inline src/shared/models/constants.ts]: weird casing

File: src/shared/pages/dashboard.tsx [hunk 2]
```diff
@@ -39,6 +39,9 @@ const Dashboard: React.FC = () => (
                             <li>
                                 <Link className="govuk-link" to="/user-search">Manage a user</Link>
                             </li>
+                            <li>
+                                <Link className="govuk-link" to="/add-user">Add a user</Link>
```
Comment (zuzak) [inline src/shared/pages/dashboard.tsx:43]:
```suggestion
                                <Link className="govuk-link" to="/add-user">Add a new user</Link>
```

File: src/shared/pages/user/addTeamToUser/__tests__/addTeamToUser.spec.tsx [hunk 3]
```diff
@@ -169,26 +178,36 @@ describe('when the submit button is clicked', () => {
...
-                value: '__userId__'
-            }, '__teamId1__');
-            expect(addUserToTeamSpy).nthCalledWith(2, {
-                label: '__user__',
-                value: '__userId__'
-            }, '__teamId2__');
-            expect(clearErrorsSpy).toBeCalled();
-        });
-    });
+    // it('should call the service and dispach actions for the selected options', async () => {
```
Comment (zuzak) [inline src/shared/pages/user/addTeamToUser/__tests__/addTeamToUser.spec.tsx]: Is this meant to be commented out? Or should this just be removed from the commit entirely?

File: src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            expect(wrapper.container).toMatchSnapshot();
+        });
+    });
+});
+
+describe('when the submit button is clicked', () => {
+    describe('and the data is filled in', () => {
+
+        beforeEach(async () => {
+            mockUser.email = 'firstname@test.com';
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx]: RFC 2606:
```suggestion
            mockUser.email = 'firstname@example.com';
```

File: src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+            fireEvent.click(submitButton);
+        });
+
+        describe('and the service call is successful', () => {
+            it('should redirect to the view user page', async () => {
+                expect.assertions(1);
+
+                await wait(() => {
+                    // history.push(`/user-view/${userUUID}`, { successMessage: constants.ADD_USER_SUCCESS });
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx]: Why is this commented out?

File: src/shared/pages/user/userAdd/addUser.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    };
+
+    return (
+        <>
+            <div className="govuk-grid-row">
+                <div className="govuk-grid-column-two-thirds-from-desktop">
+                    <Link to="/" className="govuk-back-link">Back</Link>
+                    <ErrorSummary pageError={pageError} />
+                    <h1 className="govuk-heading-xl">
+                        Add a User
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/addUser.tsx:78]:
```suggestion
                        Add a user
```

File: src/shared/pages/user/userAdd/addUser.tsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                    </h1>
+                </div>
+            </div>
+            <div className="govuk-grid-row">
+                <div className="govuk-grid-column-one-half-from-desktop">
+                    <form action="/api/user" method="POST" onSubmit={handleSubmit}>
+                        <input type="hidden" name="_csrf" value={csrfToken} />
+                        <Text
+                            label="Email Address"
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/addUser.tsx:87]:
```suggestion
                            label="Email address"
```

File: src/shared/pages/user/userAdd/addUser.tsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                            name="email"
+                            type="text"
+                            updateState={({ name, value }) => dispatch({ name, value })}
+                            value={user.email}
+                        />
+                        <Text
+                            label="First Name"
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/addUser.tsx:94]:
```suggestion
                            label="First name"
```

File: src/shared/pages/user/userAdd/addUser.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                            name="firstName"
+                            type="text"
+                            updateState={({ name, value }) => dispatch({ name, value })}
+                            value={user.firstName}
+                        />
+                        <Text
+                            label="Last Name"
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/addUser.tsx:101]:
```suggestion
                            label="Last name"
```

File: src/shared/pages/user/userAdd/addUser.tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                            name="lastName"
+                            type="text"
+                            updateState={({ name, value }) => dispatch({ name, value })}
+                            value={user.lastName}
+                        />
+                        <Submit />
```
Comment (zuzak) [inline src/shared/pages/user/userAdd/addUser.tsx:107]: Would be nice if the submit label was "Create user"

File: src/shared/pages/dashboard.tsx (hunk 2 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/dashboard.tsx:43]: This is following the existing labelling pattern in MUI.

File: src/shared/pages/user/addTeamToUser/__tests__/addTeamToUser.spec.tsx (hunk 3 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/addTeamToUser/__tests__/addTeamToUser.spec.tsx]: Ooops, no. Fixed.

File: src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx (hunk 4 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx]: Interesting! OK, fixed.

File: src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx (hunk 5 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/__tests__/addUser.spec.tsx]: I have now removed this line. I added it temporarily while working on this test.

File: src/shared/pages/user/userAdd/addUser.tsx (hunk 6 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/addUser.tsx:78]: This is following the existing labelling pattern in MUI.

File: src/shared/pages/user/userAdd/addUser.tsx (hunk 7 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/addUser.tsx:87]: This is following the existing labelling pattern in MUI.

File: src/shared/pages/user/userAdd/addUser.tsx (hunk 8 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/addUser.tsx:94]: This is following the existing labelling pattern in MUI.

File: src/shared/pages/user/userAdd/addUser.tsx (hunk 9 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/addUser.tsx:101]: This is following the existing labelling pattern in MUI.

File: src/shared/pages/user/userAdd/addUser.tsx (hunk 10 above)
Comment (SeanReedHomeOffice) [reply src/shared/pages/user/userAdd/addUser.tsx:107]: This is following the existing labelling pattern in MUI.

**Review by Paulcr-ho — APPROVED** [review]

**Pull request merged**